File: accbot-android/app/src/main/java/com/accbot/dca/exchange/ExchangeApi.kt
package com.accbot.dca.exchange

import com.accbot.dca.data.local.UserPreferences
import com.accbot.dca.domain.model.DcaResult
import com.accbot.dca.domain.model.Exchange
import com.accbot.dca.domain.model.ExchangeCredentials
import com.accbot.dca.domain.model.TradeHistoryPage
import okhttp3.OkHttpClient
import java.math.BigDecimal
import java.time.Instant
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Common interface for all exchange API implementations
 * Allows AccBot to work with multiple exchanges through unified interface
 */
interface ExchangeApi {
    val exchange: Exchange

    /**
     * Execute a market buy order
     * @param crypto Cryptocurrency symbol (e.g., "BTC")
     * @param fiat Fiat currency (e.g., "EUR")
     * @param fiatAmount Amount in fiat to spend
     * @return DcaResult with transaction details or error
     */
    suspend fun marketBuy(
        crypto: String,
        fiat: String,
        fiatAmount: BigDecimal
    ): DcaResult

    /**
     * Get current balance for a currency
     */
    suspend fun getBalance(currency: String): BigDecimal?

    /**
     * Get current market price
     */
    suspend fun getCurrentPrice(crypto: String, fiat: String): BigDecimal?

    /**
     * Withdraw crypto to external wallet
     */
    suspend fun withdraw(
        crypto: String,
        amount: BigDecimal,
        address: String
    ): Result<String> // Returns withdrawal ID or error

    /**
     * Get withdrawal fee
     */
    suspend fun getWithdrawalFee(crypto: String): BigDecimal?

    /**
     * Validate credentials (test API connection)
     */
    suspend fun validateCredentials(): Boolean

    /**
     * Get trade history for a currency pair.
     * Not all exchanges support this - default throws UnsupportedOperationException.
     * @param crypto Cryptocurrency symbol (e.g., "BTC")
     * @param fiat Fiat currency (e.g., "EUR")
     * @param sinceTimestamp Only return trades after this timestamp (for incremental import)
     * @param limit Maximum number of trades per page
     * @return A page of historical trades
     */
    suspend fun getTradeHistory(
        crypto: String,
        fiat: String,
        sinceTimestamp: Instant? = null,
        limit: Int = 100
    ): TradeHistoryPage = throw UnsupportedOperationException(
        "${exchange.displayName} does not support API trade history import"
    )
}

/**
 * Factory for creating exchange API instances.
 * Automatically configures APIs for sandbox or production mode based on user preferences.
 */
@Singleton
class ExchangeApiFactory @Inject constructor(
    private val userPreferences: UserPreferences,
    private val okHttpClient: OkHttpClient
) {
    /**
     * Create an exchange API instance.
     * @param credentials The credentials for the exchange
     * @param isSandbox Override sandbox mode (if null, uses user preference)
     */
    fun create(credentials: ExchangeCredentials, isSandbox: Boolean? = null): ExchangeApi {
        val sandboxMode = isSandbox ?: userPreferences.isSandboxMode()
        return when (credentials.exchange) {
            Exchange.COINMATE -> CoinmateApi(credentials, sandboxMode, okHttpClient)
            Exchange.BINANCE -> BinanceApi(credentials, sandboxMode, okHttpClient)
            Exchange.KRAKEN -> KrakenApi(credentials, sandboxMode, okHttpClient)
            Exchange.KUCOIN -> KuCoinApi(credentials, sandboxMode, okHttpClient)
            Exchange.BITFINEX -> BitfinexApi(credentials, sandboxMode)
            Exchange.HUOBI -> HuobiApi(credentials, sandboxMode)
            Exchange.COINBASE -> CoinbaseApi(credentials, sandboxMode)
        }
    }
}

File: accbot-ios/AccBot/Exchange/ExchangeApi.swift
import Foundation

/// Common interface for all exchange API implementations.
/// Allows AccBot to work with multiple exchanges through unified interface.
protocol ExchangeApi {
    var exchange: Exchange { get }

    /// Execute a market buy order
    func marketBuy(crypto: String, fiat: String, fiatAmount: Decimal) async -> DcaResult

    /// Get current balance for a currency
    func getBalance(currency: String) async -> Decimal?

    /// Get current market price
    func getCurrentPrice(crypto: String, fiat: String) async -> Decimal?

    /// Withdraw crypto to external wallet
    func withdraw(crypto: String, amount: Decimal, address: String) async throws -> String

    /// Get withdrawal fee
    func getWithdrawalFee(crypto: String) async -> Decimal?

    /// Validate credentials (test API connection)
    func validateCredentials() async throws -> Bool

    /// Query the status and fill details of a previously placed order
    func getOrderStatus(orderId: String) async -> Transaction?

    /// Get trade history for a currency pair
    func getTradeHistory(
        crypto: String,
        fiat: String,
        since: Date?,
        limit: Int
    ) async throws -> TradeHistoryPage
}

/// Default implementations for optional methods
extension ExchangeApi {
    func getOrderStatus(orderId: String) async -> Transaction? {
        nil
    }

    func getTradeHistory(
        crypto: String,
        fiat: String,
        since: Date? = nil,
        limit: Int = 100
    ) async throws -> TradeHistoryPage {
        throw ExchangeError.unsupportedOperation("\(exchange.displayName) does not support API trade history import")
    }
}

/// Shared helpers used by all exchange API implementations.
/// Extracted from per-class copies of parseJson/formatDecimal/roundDecimal.
extension ExchangeApi {
    func parseJson(_ data: Data) throws -> [String: Any] {
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw NetworkError.decodingError("Invalid JSON")
        }
        return json
    }

    func formatDecimal(_ value: Decimal, scale: Int) -> String {
        let handler = NSDecimalNumberHandler(
            roundingMode: .down, scale: Int16(scale),
            raiseOnExactness: false, raiseOnOverflow: false,
            raiseOnUnderflow: false, raiseOnDivideByZero: false
        )
        return NSDecimalNumber(decimal: value).rounding(accordingToBehavior: handler).stringValue
    }

    func roundDecimal(_ value: Decimal, scale: Int) -> Decimal {
        let handler = NSDecimalNumberHandler(
            roundingMode: .plain, scale: Int16(scale),
            raiseOnExactness: false, raiseOnOverflow: false,
            raiseOnUnderflow: false, raiseOnDivideByZero: false
        )
        return NSDecimalNumber(decimal: value).rounding(accordingToBehavior: handler).decimalValue
    }
}

enum ExchangeError: LocalizedError {
    case unsupportedOperation(String)
    case invalidCredentials
    case apiError(String)

    var errorDescription: String? {
        switch self {
        case .unsupportedOperation(let msg): return msg
        case .invalidCredentials: return "Invalid API credentials"
        case .apiError(let msg): return msg
        }
    }
}
